cluster/service_registry.go

```diff
@@ -17,60 +19,59 @@ type Node struct {
 }
 
 type ServiceRegistry struct {
-	kapi client.KeysAPI
+	KV clientv3.KV
 }
 
 func NewServiceRegistry(ctx context.Context, etcdAddr string) (*ServiceRegistry, error) {
-	cfg := client.Config{Endpoints: []string{etcdAddr}}
-	c, err := client.New(cfg)
+	cfg := clientv3.Config{
+		Endpoints:   []string{etcdAddr},
+		DialTimeout: 5 * time.Second,
+	}
+	c, err := clientv3.New(cfg)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create etcd client from addr %v: %w", etcdAddr, err)
 	}
 
 	return &ServiceRegistry{
-		kapi: client.NewKeysAPI(c),
+		KV: clientv3.NewKV(c),
 	}, nil
 }
 
-func (sr *ServiceRegistry) Register(ctx context.Context, serviceName, nodeName, host string, port int) error {
+func (sr *ServiceRegistry) Register(ctx context.Context, serviceName, nodeName string, host string, port int) error {
 	node := Node{Address: host, Port: port}
 	val, err := json.Marshal(node)
 	if err != nil {
 		return fmt.Errorf("failed to marshal node: %w", err)
 	}
 
 	key := filepath.Join(servicesPrefix, serviceName, nodeName)
-	if _, err = sr.kapi.Set(ctx, key, string(val), nil); err != nil {
+	if _, err = sr.KV.Put(ctx, key, string(val)); err != nil {
 		return fmt.Errorf("failed to register node: %w", err)
 	}
 
 	return nil
 }
 
-var defaultGetOptions = &client.GetOptions{
-	Recursive: true,
-	Sort:      true,
-}
-
 func (sr *ServiceRegistry) Services(ctx context.Context) (map[string][]Node, error) {
-	res, err := sr.kapi.Get(ctx, servicesPrefix, defaultGetOptions)
+	res, err := sr.KV.Get(ctx, servicesPrefix, clientv3.WithPrefix(), clientv3.WithSort(clientv3.SortByKey, clientv3.SortAscend))
 	if err != nil {
 		return nil, fmt.Errorf("failed to get services from etcd: %w", err)
 	}
 
-	serviceDir := res.Node.Nodes
-	services := make(map[string][]Node, len(serviceDir))
-	for _, service := range serviceDir {
-		nodes := service.Nodes
-		_, key := filepath.Split(service.Key)
-		services[key] = make([]Node, 0, len(nodes))
+	var node Node
+	services := make(map[string][]Node)
+	for _, Kvs := range res.Kvs {
+		if err := json.Unmarshal([]byte(Kvs.Value), &node); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal services nodes: %w", err)
+		}
 
-		for _, node := range nodes {
-			var n Node
-			if err := json.Unmarshal([]byte(node.Value), &n); err != nil {
-				return nil, fmt.Errorf("failed to unmarshal service nodes: %w", err)
+		key := strings.Split(string(Kvs.Key), "/")
```

Comment:
```suggestion
		_, key := filepath.Split(string(Kvs.Key))
```
this can get the key for you, and I don't there would be a < 1 case

Reply:
it returns `services/ping/node1`,  so the filepath split would return `services/ping` and `node1` 😢

Reply:
Oh I see, since we want `ping`->`node1`